oak_restricted_kernel/src/payload.rs

```diff
@@ -15,34 +15,55 @@
 //
 
 use crate::syscall::mmap::mmap;
+use alloc::{vec, vec::Vec};
+use anyhow::Result;
 use core::arch::asm;
-use goblin::elf64::program_header::{ProgramHeader, PF_W, PF_X, PT_LOAD};
+use goblin::elf64::program_header::{PF_W, PF_X, PT_LOAD};
+use oak_channel::Channel;
 use oak_restricted_kernel_interface::syscalls::{MmapFlags, MmapProtection};
 use x86_64::{
     structures::paging::{PageSize, Size2MiB},
     VirtAddr,
 };
 
+/// Reads a chunk of data and acknowledges the transmission by writing back the number of bytes
+/// read.
+fn read_chunk(channel: &mut dyn Channel, chunk: &mut [u8]) -> Result<()> {
+    let len: u32 = chunk
+        .len()
+        .try_into()
+        .map_err(|_| anyhow::anyhow!("chunk too big"))?;
+    channel.read(chunk)?;
+    channel.write(&len.to_be_bytes())
+}
+
+/// Reads a payload blob from the given channel.
+pub fn read_payload(channel: &mut dyn Channel) -> Result<Vec<u8>> {
```

Comment:
Is the dyn actually needed? It should be possible to reify this type?

Reply:
Oh, it actually wasn't -- at the cost of having to use `?Sized` as well.